File: c/2015/11/main.c
#include "../../util.h"

#if 0
static char *VALUE = "abcdefgh";
#else
static char *VALUE = "cqjxjnds";
#endif

bool check(char* v) {
    bool a = false;
    for (int8_t i = 0; i < strlen(v)-2; i++) {
        a |= v[i]+1 == v[i+1] && v[i+1]+1 == v[i+2];
    }
    bool b = true;
    for (int8_t i = 0; i < strlen(v); i++) {
        b &= v[i] != 'i' && v[i] != 'l' && v[i] != 'o';
    }
    char c;
    int8_t idx = 0;
    for (int8_t i = 0; i < strlen(v)-1; i++) {
        if (v[i] == v[i+1]) {
            if (idx == 0) {
                c = v[i];
                idx++;
            } else if (idx == 1 && c != v[i]) {
                idx++;
            }
        } 
    }
    return a && b && idx>1;
}

char* increment(char* v) {
    bool carry = true;
    for (int8_t i = strlen(v)-1; i >= 0; i--) {
        if (carry) {
            if (v[i] == 'z') {
                v[i] = 'a';
            } else {
                v[i] = v[i]+1;
                carry = false;
            }
        } else break;
    }
    // printf(">%s\n",v);
    return v;
}
 
char* next(char* v) {
    while (!check(v)) {
        v = increment(v);
    }
    return v;
}


void part1() {
    char c[9];
    strcpy(c,VALUE);
    char* cc = next(c);
    printf("Part 1: %s\n", cc);
}

void part2() {
    char c[9];
    strcpy(c,VALUE);
    char* cc = next(c);
    cc = increment(cc);
    cc = next(cc);
    printf("Part 2: %s\n", cc);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2015/11/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def check(n):
    a = False
    for i in range(len(n)-2):
        a |= ord(n[i])+1 == ord(n[i+1]) and ord(n[i+1])+1 == ord(n[i+2])
    b = not ('i' in n or 'l' in n or 'o' in n)
    c = []
    for i in range(len(n)-1):
        if n[i] == n[i+1]:
            c.append(n[i])
    return a and b and len(set(c)) == 2

def increment(n):
    m = []
    over = True
    for i in range(len(n)-1, -1, -1):
        if over:
            if n[i] == 'z':
                m.append('a')
            else:
                m.append(chr(ord(n[i])+1))
                over = False
        else:
            m.append(n[i])
    m.reverse()
    n = "".join(m)
    return n

def next(n):
    while not check(n):
        n = increment(n)
    return n

def read_input():
    return "abcdefgh" if debug else "cqjxjnds"

def part1():
    A = read_input()
    k = next(A)
    print("Part 1:",k)

def part2():
    A = read_input()
    k = increment(next(A))
    k = next(k)
    print("Part 2:",k)

part1()
part2()
